Add Async Factory Support

nadege opened the pull request:
Hello,

I heard you may be interested in a PR to add support for async Factories, related issue :  #679

We use the solution I documented in the issue in a project at work (`aiohttp` + `aiopg`) and it works well,
but even if we have models with some complexity (lot of `Subfactory` themselves using `SelfAttribute`),
I know we don't use a lot of features of Factory Boy...

For this PR I chose to add a new strategy for async creation, and added a new class `AsyncFactory` that default to this strategy.
I also added a few methods to `class Factory`, and functions to the `helpers` module,
to create, batch create, generate, batch generate... with this new strategy.
There should be no issue with retro-compatibility, it only add new features.

I wrote some tests and documentation but they may be insufficient.
Advice on what to test or document further is welcome of course.

Duplicating all tests to run them with an async factory seems a bit overkill to me, 
but maybe we should for some test cases ?

what do you think ?

francoisfreitag commented on factory/base.py:
```diff
@@ -504,14 +509,46 @@ def _create(cls, model_class, *args, **kwargs):
+            return await cls._create_model_async(model_class, *args, **kwargs)
+
+        # A Task can be awaited multiple times, unlike a coroutine.
+        # useful when a factory and a subfactory must share a same object
```
```suggestion
        # Useful when a factory and a subfactory must share the same object.
```

francoisfreitag reviewed (commented):
Hi @nadege,

Thanks a lot for taking the time to prepare this patch :tada:.

Async capabilities in factory_boy seems to be interesting to a portion of the user base and would benefit the library. Django is planning to eventually [support creating objects asynchronously](https://github.com/django/deps/blob/master/accepted/0009-async.rst#the-orm) and SQLAlchemy has [GINO](https://github.com/fantix/gino) or `aiopg`, movement in that direction is great.

Thanks for adding tests following the create methods, it helped review a lot. The suggested changes make sense to me, though I am not very familiar with async in Python.

---

<s>Can tests be run with the [Python development mode](https://docs.python.org/3.9/library/asyncio-dev.html#debug-mode) to catch potential issues in asyncio?</s>
Enabled in https://github.com/FactoryBoy/factory_boy/pull/812.

The new extension points should be listed in the extension points section https://github.com/FactoryBoy/factory_boy/blob/9b256e48dd58ef7c97e3ddf09d972d1bb1d2df8e/docs/reference.rst.

An entry should be added to the release note.

nadege commented on tests/test_using.py:
```diff
@@ -556,6 +669,36 @@ class Meta:
         test_model = TestModel2Factory()
         self.assertEqual(4, test_model.two.three)
 
+    async def test_self_attribute_parent_async(self):
+        class AsyncTestModel2(FakeAsyncModel):
+            pass
+
+        class AsyncTestModel3(FakeAsyncModel):
+            pass
+
+        class TestAsyncModel3Factory(FakeAsyncModelFactory):
+            class Meta:
+                model = AsyncTestModel3
+            toto = 4
+
+        class TestAsyncModelFactory(FakeAsyncModelFactory):
+            class Meta:
+                model = AsyncTestModel
+            one = 3
+            three = factory.SelfAttribute('..bar')
+
+        class TestAsyncModel2Factory(FakeAsyncModelFactory):
+            class Meta:
+                model = AsyncTestModel2
+            zero = 4
+            two = factory.SubFactory(TestAsyncModelFactory, one=1)
+            bar = factory.SubFactory(TestAsyncModel3Factory, toto=5, tata=factory.SelfAttribute('..zero'))
+
+        test_model = await TestAsyncModel2Factory()
+        self.assertEqual(test_model.bar, test_model.two.three)
+        self.assertEqual(5, test_model.two.three.toto)
+        self.assertEqual(4, test_model.bar.tata)
```
I agree, I try to keep "local coding/testing style" but you're right this would be way better.

nadege commented on docs/introduction.rst:
```diff
@@ -351,13 +352,64 @@ Calling a :class:`~factory.Factory` subclass will provide an object through the
+If your project uses asynchronous code to save objects to database, you
+can use that in your factories too.
+
+You need to override the async method :meth:`factory.Factory._create_model_async`
+to define how your objects are created and saved to the database.
```
Yes I'd like to do that. 
I don't know which library to choose though, I will look into it but suggestions are welcome.

francoisfreitag replied:
IMO, you should go with the one you’re most familiar with. Perhaps a stripped down version of what you are using on projects at work (built on `aiopg`)?

Other contributors can then follow the lead and build support for their favorite library.

nadege commented:
Thanks for the review!
I'll fix the simpler things first (doc, tests..) and then try to add an implementation of `_create_model_async` for a  library.

nadege commented on factory/helpers.py:
```diff
@@ -62,6 +62,24 @@ def create_batch(klass, size, **kwargs):
+    if "FACTORY_CLASS" not in kwargs:
+        kwargs["FACTORY_CLASS"] = base.AsyncFactory
+    assert issubclass(kwargs["FACTORY_CLASS"], base.AsyncFactory), \
+        "FACTORY_CLASS must be a subclass of AsyncFactory when using create_async_batch"
```
yes, I removed the constraint

nadege commented:
I added an implementation for https://github.com/encode/orm ORM in the last commit.
It's the better I found which support async and sqlite.
I don't know how to simply add a postgresql database to the current CI, 
but when it'll be replaced by github actions it should be easier.
 I'll work on an implementation for a PG backed ORM then.

Last problem: `unittest.IsolatedAsyncioTestCase` is new in python3.8
I need to find a solution for previous python versions.

(I also have documentation spelling errors.)

nadege commented:
I resolved the compatibility issues with older python versions and spelling errors in the doc. 
Should be ready for a new review

nadege commented:
oh I still need to update `/docs/reference.rst`,  the release note and CREDIT files.

nadege commented:
I'll wait for djangoproject.com to be fixed so I can run `make doc` before pushing my next commit. 
Also that's why CI failed on my last commit.

francoisfreitag reviewed (commented):
I took a quick look at it today, it seems great! :tada:
I’ll try to dive deeper later this week. Thanks for adding support for SQLite, it helps exercising the new code and increases confidence in the patch.

---

> I don't know how to simply add a postgresql database to the current CI, but when it'll be replaced by github actions it should be easier.

https://docs.travis-ci.com/user/database-setup/#postgresql should be helpful. If that does not work out, I’m happy with the SQLite version to start with. The more test coverage, the better, but IMO PostgreSQL support is not mandatory for landing this patch. Getting started on async benefits the project and we can increase test coverage as we go.

francoisfreitag commented on tests/test_using.py:
```diff
@@ -555,6 +677,50 @@ class Meta:
+        self.assertEqual("Toronto", book.author.hometown)
+        self.assertEqual(2, book.chapter.id)
+        self.assertEqual(2, book.preface.id)
+        self.assertEqual(2, book.preface.author.id)
```
This is not meaningful. The preface author would actually be a different object in the DB and have a different id from the book author. One is from Paris, the other from Toronto.

IIUC, the ids are checked to make sure the `create()` method of `FakeAsyncModel` has been called? If so, thoughts on adding a comment and/or clarifying the magic value with a unique object for `id`? For example:
```python
create_marker = object()

class FakeAsyncModel:
    @classmethod
    async def create(cls, **kwargs):
        instance = cls(**kwargs)
        if not instance.id:
            instance.id = create_marker
        # ...

        # In the tests:
        self.assertIs(book.author.id, create_marker)
```
?

nadege replied:
> IIUC, the ids are checked to make sure the create() method of FakeAsyncModel has been called?
exact.
I'll improve that

nadege commented on tox.ini:
```diff
@@ -19,6 +19,8 @@ deps =
     django{22,30,31}: Pillow
     alchemy: SQLAlchemy
     mongoengine: mongoengine
+    orm: orm~=0.1
```
It's advised by `orm` maintainers.
from their doc: 

> ORM is still under development: We recommend pinning any dependencies with orm~=0.1

francoisfreitag replied:
Works for me, thanks!

nadege commented on tests/test_orm.py:
```diff
@@ -0,0 +1,42 @@
+        loop = asyncio.get_event_loop()
+
+        database_context = models.database
+        loop.run_until_complete(database_context.__aenter__())
```
oh yes I like that solution.

nadege commented:
Issues highlighted in previous review should be fixed. Added doc, credits and changelog. 
Ready for next review :)

francoisfreitag commented on docs/introduction.rst:
```diff
@@ -351,13 +352,52 @@ Calling a :class:`~factory.Factory` subclass will provide an object through the
+    <MyClass: X (unsaved)>
+
+    >>> MyFactory()  # equivalent to MyFactory.create_async()
+    <MyClass: X (saved using async code)>
```
Missing a new line at the end of the file.

francoisfreitag commented on tests/test_using.py:
```diff
@@ -171,6 +212,46 @@ def test_create_batch_custom_base(self):
+    def test_create_async(self):
+        loop = asyncio.get_event_loop()
+        obj = loop.run_until_complete(factory.create_async(FakeAsyncModel, foo='bar'))
+        self.assertEqual(obj.id, None)
```
`id` is `None` because `create()` was not called.

Is that the expected behavior? I would expect that when the task to create the object is done, the object has been recorded to the database and has an id?

nadege replied:
It's expected because the `make_factory` function will output a factory based on Factory class.
So we end up with model_class(**kwargs) to create the object,  not the model async `create` function.
I'll add comments to the test.

For comparison, the sync test:
```
    def test_create(self):
        obj = factory.create(FakeModel, foo='bar')
        self.assertEqual(obj.id, None)
        self.assertEqual(obj.foo, 'bar')
```

If you look at the test `test_create_async_custom_base`, it has parameter `FACTORY_CLASS=FakeAsyncModelFactory` and the async creation is ok.

francoisfreitag reviewed (commented):
I wonder how async factories relate to post generation hooks https://factoryboy.readthedocs.io/en/latest/reference.html#postgeneration?

The post generation hooks assume the instance was created by the call to instantiate:

https://github.com/FactoryBoy/factory_boy/blob/88b81ead8e56a8c3eef268c299616fa3a958f091/factory/builder.py#L268-L272

Since `create_async` returns a task and not an instance, perhaps the post generation hooks should await it before to run? Or register the post generation hooks as `add_done_callback` on the Task?
Or do you expect subclasses of AsyncFactory to expect the task as the first positional argument, awaiting it if necessary?

TheLazzziest commented:
Hi, @nadege . Great work you've done. However, I would like to ask will you plan to add async support for post_generation methods ? I haven't found any code changes related to that feature.

nadege commented:
Hello ! 
Ok I found time to continue work on this. I'll look into the post generation feature. I think I never used that  before.

nadege commented:
I handled the remarks you made in the last review. Now I'll look into post-generation

mdczaplicki commented:
Seems that `LazyAttribute` that depends on an async `SubFactory` receives a `Task` instead of an object.

nadege commented:
@mdczaplicki would you have a test that reproduce the issue ?

mdczaplicki commented:
Sure, I'll provide you with something next week. :)

On Jan 22 2021, at 11:06 am, Nadege Michel <notifications@github.com> wrote:
>
>
> @mdczaplicki (https://github.com/mdczaplicki) would you have a test that reproduce the issue ?
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub (https://github.com/FactoryBoy/factory_boy/pull/803#issuecomment-765293517), or unsubscribe (https://github.com/notifications/unsubscribe-auth/ACFPY2QYTE2IMOMIRBU7NE3S3FE2DANCNFSM4TFHREYQ).
>

fadedDexofan commented:
@nadege @mdczaplicki Any progress on this?

nadege commented:
No, I'm between jobs and my personal computer just broke, so I can't work on this at the moment.

romulorosa commented:
Hey @nadege the work here is awesome! 🥇 I am really interested on having this feature kicked out of the door, do you have any plans to finish it? If you don't mind I would be happy to jump in and help to get it done. Could you tell what is missing?

francoisfreitag commented:
IIRC, the main piece remaining is deciding how to handle the [post-generation hooks](https://factoryboy.readthedocs.io/en/latest/reference.html#post-generation-hooks).

nadege commented:
hi @romulorosa ! I don't have much time (or energy :x) to continue this short term. So I'd be happy to see you take it from here !

gegnew commented:
any updates on this PR?

ShipraShalini commented:
> IIRC, the main piece remaining is deciding how to handle the [post-generation hooks](https://factoryboy.readthedocs.io/en/latest/reference.html#post-generation-hooks).

@francoisfreitag Can this feature be released in parts? As in, first release the basic functionality and then add the `post-generation-hooks` in a future release.

francoisfreitag commented:
The post generation hooks need handling. The handling can probably be to `raise NotImplemented` when a post generation hook is declared on an `AsyncFactory`. Although clearly not ideal, that would at least improve the use case for factory boy in an async context.

ShipraShalini commented:
My use case doesn't require `post-generation-hooks`, so I can work on getting it out with `raise NotImplemented`.

coneybeare commented:
I'd like to see this PR get merged too. @ShipraShalini did you make some progress there on the `raise NotImplemented` ? Can I offer assistance?

nadege commented:
Hello there, I'm back 🙇 

Changes on the PR since last time (3 years ago, time flies... 👵)
- 😅 Rebase 
- 🔥 Drop python 3.6 support.
- 🔄  Replaced [orm](https://github.com/encode/orm) implementation by a SQLAlchemy one.
- 🧪 One test for Post Generation hook. (It just worked!)


TODO:
- [ ] 🛠️  Fix lint & other CI failures 
- [ ] 🧪 More tests? (please advise!)

nadege commented on tests/test_using.py:
```diff
@@ -2248,6 +2521,29 @@ def incr_one(self, _create, _increment):
+            def incr_one(self, _create, _increment):
+                self.one += 1
+
+        obj = TestAsyncFactory.create()
```
This test is incorrect. This is not using async creation.  😅 
That was too good to be true... 😁

nadege commented:
And, post generation is now handled too! 🎉 

I played with the suggestion from @francoisfreitag, to use `add_done_callback`.
Now the whole post generation step is a callback when dealing with a Task.

Code is in last commit, it's a bit quick & dirty. Not sure of the cleanest way to declare the `_handle_post_generation` function.
Happy to make any change or add more tests. Please advise 🙇

mlorenc-sonnen commented:
Great work @nadege , it was very helpful in fixing some of the issues with the solution in my team. I also hope this gets merged soon so we can migrate away from temporary workarounds :)

nextmat commented:
Appreciate all the work on this @nadege. :-) 

@francoisfreitag is there other work needed to get this merged? Thanks!

francoisfreitag reviewed (commented):
Great work overall, thank you!! :star: :raised_hands:

I believe this is ready for merging, pending some minor RST fixes and maybe a couple of test tightening. I’ll try to push a few commits in the upcoming weeks, depending on my newborn’s sleep.
Feel free to beat me to it, if you have time before then!

Pinging @rbarrois, in case you would like to review this work before inclusion.

nadege commented on tests/test_alchemy.py:
```diff
@@ -337,3 +338,61 @@ class Meta:
+        def setUp(self):
+            super().setUp()
+
+            async def asyncSetUp():
+                async with models.async_engine.begin() as conn:
+                    await conn.run_sync(models.Base.metadata.drop_all)
+                    await conn.run_sync(models.Base.metadata.create_all)
+
+                NoteFactory.reset_sequence(0)
+
+            asyncio.run(asyncSetUp())
```
> New in version 3.8.

Haha! I actually used it in an [earlier version](https://github.com/FactoryBoy/factory_boy/compare/5f0c9f6c30eeb32baefa8518c2d025bbeb6505e2..330c5e9cfec5253bc9adb90cd0fc5268362a716a), but at that time FactoryBoy was still supporting earlier python versions so I dropped it. 
I'll add it back :)

nadege commented on factory/alchemy.py:
```diff
@@ -140,8 +140,21 @@ class Meta:
+            await session.flush()
+        elif session_persistence == SESSION_PERSISTENCE_COMMIT:
+            await session.commit()
+        return obj
```
Code was not actually saving the object to the db. 😬 

TODO:
- add a test that would have shown that
- handle `get_or_create` in async create as well, or can we ship without? (I won't have bandwidth for it)

nadege commented:
I rebased & addressed the simpler comments, realised we were not committing/flushing so I fixed that.

TODO:
- [ ] test that we actually save objects in slqalchemy implementation
- [ ] use [IsolatedAsyncioTestCase](https://docs.python.org/3/library/unittest.html#unittest.IsolatedAsyncioTestCase) in tests
- [ ] handle `get_or_create` in async implem for sqlalchemy?

I can find time for first 2 items.

rushilsrivastava commented:
Hey @nadege! How can I help move this PR along? I would love to see this merged in :)

I am happy to tackle the `get_or_create` async implementation

Adman commented:
@rushilsrivastava @nadege Maybe the `get_or_create` could be inspired by @ADR-007's PR in async-factory-boy: https://github.com/kuzxnia/async_factory_boy/pull/8/files#diff-92489e997002e4b4508e74d518bb8984a51a91c12c466c661772a6878629b31f
